core/src/main/java/com/linecorp/armeria/server/ShutdownSupport.java

```diff
@@ -35,31 +35,48 @@ static ShutdownSupport of(AccessLogWriter accessLogWriter) {
         });
     }
 
-    static ShutdownSupport of(ScheduledExecutorService executor) {
+    static ShutdownSupport of(ExecutorService executor) {
         requireNonNull(executor, "executor");
         return () -> {
-            final ScheduledExecutorService e;
-            if (executor instanceof UnstoppableScheduledExecutorService) {
-                e = ((UnstoppableScheduledExecutorService) executor).getExecutorService();
-            } else {
-                e = executor;
-            }
-
-            e.shutdown();
-
+            executor.shutdown();
             boolean interrupted = false;
             try {
-                while (!e.isTerminated()) {
+                while (!executor.isTerminated()) {
                     try {
-                        e.awaitTermination(1, TimeUnit.HOURS);
+                        executor.awaitTermination(1, TimeUnit.HOURS);
                     } catch (InterruptedException ignore) {
                         interrupted = true;
                     }
                 }
             } catch (Exception cause) {
-                logger.warn("Failed to shutdown the {}:", e, cause);
+                logger.warn("Failed to shutdown the {}:", executor, cause);
+            } finally {
+                if (interrupted) {
+                    Thread.currentThread().interrupt();
+                }
             }
-            if (interrupted) {
+            return UnmodifiableFuture.completedFuture(null);
+        };
+    }
+
+    static ShutdownSupport of(ExecutorService executor, long terminationTimeoutMillis) {
+        requireNonNull(executor, "executor");
+        return () -> {
+            executor.shutdown();
+            try {
+                if (!executor.awaitTermination(terminationTimeoutMillis, TimeUnit.MILLISECONDS)) {
+                    logger.warn("As the termination does not complete within the specified timeout, an " +
```

Comment:
How about
`logger.warn("Failed to terminate {} in {} millis. Shutdown it forcefully.", executor, terminationTimeoutMillis);`